[Feature]: option to exclude tests which require network connectivity

### What would you like to see added to the NWBInspector?

I get a lot of errors of the form:

```
E   OSError: Unable to open file (curl cannot perform request)
```

When trying to run the test suite of versions `0.4.{3..12}` this being the full build log for the latter: https://ppb.chymera.eu/ec7618.log

This could be done by either having an option which just turns network-reliant tests off (e.g. depending on an environment variable [as seen in DANDI](https://github.com/dandi/dandi-cli/blob/17e0b04263ed48b3c2f5946b20daded49bef0dfa/dandi/tests/skip.py#L131)), or by releasing static test data which can be fetched explicitly instead of implicitly (which is the reason why tests often are and generally should be network-sandboxed). This latter option is used by ITK (not the additional release links):

![decohost_2022-09-06T15:05:49](https://user-images.githubusercontent.com/950524/188718515-0a706f35-50df-4083-902e-03465f02ffc1.png)

The band-aid solution is [to simply disable the affected tests manually `pytest ... --ignore tests/test_inspector.py --ignore tests/unit_tests`](https://ppb.chymera.eu/1b2995.log), but this is likely unsustainable as things change in future releases, so it would be cool to have something more reliable.

### Do you have any interest in helping implement the feature?

No.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/nwbinspector/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this feature was not already [requested](https://github.com/neurodatawithoutborders/nwbinspector/issues)?

--- a/docs/developer_guide.rst
+++ b/docs/developer_guide.rst
@@ -35,3 +35,17 @@ Then, all that is needed for this to be automatically included when you run the 
 the modules flag ``-m`` or ``--modules`` along with the name of your module that contains the custom check. If using
 the library instead, you need only import the ``available_checks`` global variable from your own submodules, or
 otherwise import your check functions after importing the ``nwbinspector`` in your ``__init__.py``.
+
+
+Disable Tests That Require Network Connection
+---------------------------------------------
+
+Some of the tests in the suite require internet connectivity both to and from the DANDI archive S3 bucket.
+If this is failing for some reason, you can explicitly control all related tests by setting the environment variable
+``DANDI_TESTS_NONETWORK`` to some value able to be parsed by ``distutils.util.str2tool``. For example, to disable them on
+a linux system, run
+
+.. code-block::
+    export DANDI_TESTS_NONETWORK=1
+    
+in your environment before running ``pytest``.

--- a/docs/developer_guide.rst
+++ b/docs/developer_guide.rst
@@ -42,10 +42,10 @@ Disable Tests That Require Network Connection
 
 Some of the tests in the suite require internet connectivity both to and from the DANDI archive S3 bucket.
 If this is failing for some reason, you can explicitly control all related tests by setting the environment variable
-``DANDI_TESTS_NONETWORK`` to some value able to be parsed by ``distutils.util.str2tool``. For example, to disable them on
+``NWBI_SKIP_NETWORK_TESTS`` to some value able to be parsed by ``distutils.util.str2tool``. For example, to disable them on
 a linux system, run
 
 .. code-block::
-    export DANDI_TESTS_NONETWORK=1
+    export NWBI_SKIP_NETWORK_TESTS=1
 
 in your environment before running ``pytest``.

--- a/docs/developer_guide.rst
+++ b/docs/developer_guide.rst
@@ -47,5 +47,5 @@ a linux system, run
 
 .. code-block::
     export DANDI_TESTS_NONETWORK=1
-    
+
 in your environment before running ``pytest``.

--- a/nwbinspector/testing.py
+++ b/nwbinspector/testing.py
@@ -1,0 +1,35 @@
+"""Helper functions for internal use across the testing suite."""
+import os
+from distutils.util import strtobool
+from typing import Tuple, Optional
+
+from .tools import check_streaming_enabled
+from .utils import is_module_installed
+
+
+def check_streaming_tests_enabled() -> Tuple[bool, Optional[str]]:
+    """
+    General purpose helper for determining if the testing environment can support S3 DANDI streaming.
+
+    Returns the boolean status of the check and, if False, provides a string reason for the failure for the user to
+    utilize as they please (raise an error or warning with that message, print it, or ignore it).
+    """
+    failure_reason = ""
+
+    environment_skip_flag = os.environ.get("NWBI_SKIP_NETWORK_TESTS", "")
+    environment_skip_flag_bool = (
+        strtobool(os.environ.get("NWBI_SKIP_NETWORK_TESTS", "")) if environment_skip_flag != "" else False
+    )
+    if environment_skip_flag_bool:
+        failure_reason += "Environmental variable set to skip network tets."
+
+    streaming_enabled, streaming_failure_reason = check_streaming_enabled()
+    if not streaming_enabled:
+        failure_reason += streaming_failure_reason
+
+    have_dandi = is_module_installed("dandi")
+    if not have_dandi:
+        failure_reason += "The DANDI package is not installed on the system."
+
+    failure_reason = None if failure_reason == "" else failure_reason
+    return streaming_enabled and not environment_skip_flag_bool and have_dandi, failure_reason

--- a/nwbinspector/tools.py
+++ b/nwbinspector/tools.py
@@ -2,9 +2,12 @@
 import re
 from uuid import uuid4
 from datetime import datetime
-from typing import Optional, Dict
+from typing import Optional, Dict, Tuple
 from concurrent.futures import ProcessPoolExecutor, as_completed
+from urllib import request
+from warnings import warn
 
+import h5py
 from pynwb import NWBFile
 
 from .utils import is_module_installed, calculate_number_of_cpu
@@ -75,3 +78,19 @@ def _get_content_url_and_path(asset, follow_redirects: int = 1, strip_query: boo
     Must be globally defined (not as a part of get_s3_urls..) in order to be pickled.
     """
     return {asset.get_content_url(follow_redirects=1, strip_query=True): asset.path}
+
+
+def check_streaming_enabled() -> Tuple[bool, Optional[str]]:
+    """
+    General purpose helper for determining if the environment can support S3 DANDI streaming.
+
+    Returns the boolean status of the check and, if False, provides a string reason for the failure for the user to
+    utilize as they please (raise an error or warning with that message, print it, or ignore it).
+    """
+    try:
+        request.urlopen("https://dandiarchive.s3.amazonaws.com/ros3test.nwb", timeout=1)
+    except request.URLError:
+        return False, "Internet access to DANDI failed."
+    if "ros3" not in h5py.registered_drivers():
+        return False, "ROS3 driver not installed."
+    return True, None

--- a/tests/test_inspector.py
+++ b/tests/test_inspector.py
@@ -5,7 +5,6 @@ from tempfile import mkdtemp
 from pathlib import Path
 from unittest import TestCase
 from datetime import datetime
-from distutils.util import strtobool
 
 import numpy as np
 from pynwb import NWBFile, NWBHDF5IO, TimeSeries
@@ -25,25 +24,11 @@ from nwbinspector import (
 )
 from nwbinspector import inspect_all, inspect_nwb
 from nwbinspector.register_checks import Severity, InspectorMessage, register_check
-from nwbinspector.utils import FilePathType, is_module_installed
+from nwbinspector.testing import check_streaming_tests_enabled
 from nwbinspector.tools import make_minimal_nwbfile
+from nwbinspector.utils import FilePathType
 
-DANDI_TESTS_NONETWORK = os.environ.get("DANDI_TESTS_NONETWORK", "")
-NO_NETWORK = strtobool(DANDI_TESTS_NONETWORK) if DANDI_TESTS_NONETWORK != "" else False
-if not NO_NETWORK:
-    try:
-        with NWBHDF5IO(
-            path="https://dandiarchive.s3.amazonaws.com/blobs/11e/c89/11ec8933-1456-4942-922b-94e5878bb991",
-            mode="r",
-            load_namespaces=True,
-            driver="ros3",
-        ) as io:
-            nwbfile = io.read()
-        HAVE_ROS3 = True
-    except ValueError:  # ValueError: h5py was built without ROS3 support, can't use ros3 driver
-        HAVE_ROS3 = False
-HAVE_DANDI = is_module_installed("dandi")
-DISABLE_STREAMING_TESTS = NO_NETWORK or not HAVE_ROS3 or not HAVE_DANDI
+DISABLE_STREAMING_TESTS, DISABLE_STREAMING_TESTS_REASON = check_streaming_tests_enabled()
 
 
 def add_big_dataset_no_compression(nwbfile: NWBFile):
@@ -603,7 +588,7 @@ def test_dandiset_streaming_parallel():
     )
 
 
-@pytest.mark.skipif(DISABLE_STREAMING_TESTS, reason="Needs h5py setup with ROS3.")
+@pytest.mark.skipif(DISABLE_STREAMING_TESTS, reason=DISABLE_STREAMING_TESTS_REASON or "")
 class TestStreamingCLI(TestCase):
     @classmethod
     def setUpClass(cls):

--- a/tests/test_inspector.py
+++ b/tests/test_inspector.py
@@ -5,6 +5,7 @@ from tempfile import mkdtemp
 from pathlib import Path
 from unittest import TestCase
 from datetime import datetime
+from distutils.util import strtobool
 
 import numpy as np
 from pynwb import NWBFile, NWBHDF5IO, TimeSeries
@@ -27,19 +28,22 @@ from nwbinspector.register_checks import Severity, InspectorMessage, register_ch
 from nwbinspector.utils import FilePathType, is_module_installed
 from nwbinspector.tools import make_minimal_nwbfile
 
-
-try:
-    with NWBHDF5IO(
-        path="https://dandiarchive.s3.amazonaws.com/blobs/11e/c89/11ec8933-1456-4942-922b-94e5878bb991",
-        mode="r",
-        load_namespaces=True,
-        driver="ros3",
-    ) as io:
-        nwbfile = io.read()
-    HAVE_ROS3 = True
-except ValueError:  # ValueError: h5py was built without ROS3 support, can't use ros3 driver
-    HAVE_ROS3 = False
+DANDI_TESTS_NONETWORK = os.environ.get("DANDI_TESTS_NONETWORK", "")
+NO_NETWORK = strtobool(DANDI_TESTS_NONETWORK) if DANDI_TESTS_NONETWORK != "" else False
+if not NO_NETWORK:
+    try:
+        with NWBHDF5IO(
+            path="https://dandiarchive.s3.amazonaws.com/blobs/11e/c89/11ec8933-1456-4942-922b-94e5878bb991",
+            mode="r",
+            load_namespaces=True,
+            driver="ros3",
+        ) as io:
+            nwbfile = io.read()
+        HAVE_ROS3 = True
+    except ValueError:  # ValueError: h5py was built without ROS3 support, can't use ros3 driver
+        HAVE_ROS3 = False
 HAVE_DANDI = is_module_installed("dandi")
+DISABLE_STREAMING_TESTS = NO_NETWORK or not HAVE_ROS3 or not HAVE_DANDI
 
 
 def add_big_dataset_no_compression(nwbfile: NWBFile):
@@ -571,7 +575,7 @@ class TestInspector(TestCase):
         self.assertCountEqual(first=test_results, second=true_results)
 
 
-@pytest.mark.skipif(not HAVE_ROS3 or not HAVE_DANDI, reason="Needs h5py setup with ROS3.")
+@pytest.mark.skipif(DISABLE_STREAMING_TESTS, reason="Needs h5py setup with ROS3.")
 def test_dandiset_streaming():
     messages = list(inspect_all(path="000126", select=["check_subject_species_exists"], stream=True))
     assert messages[0] == InspectorMessage(
@@ -585,7 +589,7 @@ def test_dandiset_streaming():
     )
 
 
-@pytest.mark.skipif(not HAVE_ROS3 or not HAVE_DANDI, reason="Needs h5py setup with ROS3.")
+@pytest.mark.skipif(DISABLE_STREAMING_TESTS, reason="Needs h5py setup with ROS3.")
 def test_dandiset_streaming_parallel():
     messages = list(inspect_all(path="000126", select=["check_subject_species_exists"], stream=True, n_jobs=2))
     assert messages[0] == InspectorMessage(
@@ -599,7 +603,7 @@ def test_dandiset_streaming_parallel():
     )
 
 
-@pytest.mark.skipif(not HAVE_ROS3 or not HAVE_DANDI, reason="Needs h5py setup with ROS3.")
+@pytest.mark.skipif(DISABLE_STREAMING_TESTS, reason="Needs h5py setup with ROS3.")
 class TestStreamingCLI(TestCase):
     @classmethod
     def setUpClass(cls):

--- a/tests/unit_tests/test_time_series.py
+++ b/tests/unit_tests/test_time_series.py
@@ -1,6 +1,3 @@
-import os
-from distutils.util import strtobool
-
 import numpy as np
 import pynwb
 import pytest
@@ -16,25 +13,10 @@ from nwbinspector import (
     check_missing_unit,
     check_resolution,
 )
-from nwbinspector.utils import get_package_version, robust_s3_read, is_module_installed
+from nwbinspector.testing import check_streaming_tests_enabled
+from nwbinspector.utils import get_package_version, robust_s3_read
 
-DANDI_TESTS_NONETWORK = os.environ.get("DANDI_TESTS_NONETWORK", "")
-NO_NETWORK = strtobool(DANDI_TESTS_NONETWORK) if DANDI_TESTS_NONETWORK != "" else False
-if not NO_NETWORK:
-    try:
-        # Test ros3 on sub-YutaMouse54/sub-YutaMouse54_ses-YutaMouse54-160630_behavior+ecephys.nwb from #3
-        with pynwb.NWBHDF5IO(
-            path="https://dandiarchive.s3.amazonaws.com/blobs/f03/18e/f0318e30-4f4f-466d-a8e9-a962863e3081",
-            mode="r",
-            load_namespaces=True,
-            driver="ros3",
-        ) as io:
-            nwbfile = io.read()
-        HAVE_ROS3 = True
-    except ValueError:  # ValueError: h5py was built without ROS3 support, can't use ros3 driver
-        HAVE_ROS3 = False
-HAVE_DANDI = is_module_installed("dandi")
-DISABLE_STREAMING_TESTS = NO_NETWORK or not HAVE_ROS3 or not HAVE_DANDI
+DISABLE_STREAMING_TESTS, DISABLE_STREAMING_TESTS_REASON = check_streaming_tests_enabled()
 
 
 def test_check_regular_timestamps():
@@ -186,7 +168,7 @@ def test_check_unknown_resolution_pass():
 
 @pytest.mark.skipif(
     DISABLE_STREAMING_TESTS or get_package_version("hdmf") >= version.parse("3.3.1"),
-    reason="Needs h5py setup with ROS3, as well as 'hdmf<3.3.1'.",
+    reason=f"{DISABLE_STREAMING_TESTS_REASON}. Also needs 'hdmf<3.3.1'.",
 )
 def test_check_none_matnwb_resolution_pass():
     """

--- a/tests/unit_tests/test_time_series.py
+++ b/tests/unit_tests/test_time_series.py
@@ -1,9 +1,10 @@
-from packaging import version
-from time import sleep
+import os
+from distutils.util import strtobool
 
 import numpy as np
 import pynwb
 import pytest
+from packaging import version
 
 from nwbinspector import (
     InspectorMessage,
@@ -15,20 +16,25 @@ from nwbinspector import (
     check_missing_unit,
     check_resolution,
 )
-from nwbinspector.utils import get_package_version, robust_s3_read
+from nwbinspector.utils import get_package_version, robust_s3_read, is_module_installed
 
-try:
-    # Test ros3 on sub-YutaMouse54/sub-YutaMouse54_ses-YutaMouse54-160630_behavior+ecephys.nwb from #3
-    with pynwb.NWBHDF5IO(
-        path="https://dandiarchive.s3.amazonaws.com/blobs/f03/18e/f0318e30-4f4f-466d-a8e9-a962863e3081",
-        mode="r",
-        load_namespaces=True,
-        driver="ros3",
-    ) as io:
-        nwbfile = io.read()
-    HAVE_ROS3 = True
-except ValueError:  # ValueError: h5py was built without ROS3 support, can't use ros3 driver
-    HAVE_ROS3 = False
+DANDI_TESTS_NONETWORK = os.environ.get("DANDI_TESTS_NONETWORK", "")
+NO_NETWORK = strtobool(DANDI_TESTS_NONETWORK) if DANDI_TESTS_NONETWORK != "" else False
+if not NO_NETWORK:
+    try:
+        # Test ros3 on sub-YutaMouse54/sub-YutaMouse54_ses-YutaMouse54-160630_behavior+ecephys.nwb from #3
+        with pynwb.NWBHDF5IO(
+            path="https://dandiarchive.s3.amazonaws.com/blobs/f03/18e/f0318e30-4f4f-466d-a8e9-a962863e3081",
+            mode="r",
+            load_namespaces=True,
+            driver="ros3",
+        ) as io:
+            nwbfile = io.read()
+        HAVE_ROS3 = True
+    except ValueError:  # ValueError: h5py was built without ROS3 support, can't use ros3 driver
+        HAVE_ROS3 = False
+HAVE_DANDI = is_module_installed("dandi")
+DISABLE_STREAMING_TESTS = NO_NETWORK or not HAVE_ROS3 or not HAVE_DANDI
 
 
 def test_check_regular_timestamps():
@@ -179,7 +185,7 @@ def test_check_unknown_resolution_pass():
 
 
 @pytest.mark.skipif(
-    not HAVE_ROS3 or get_package_version("hdmf") >= version.parse("3.3.1"),
+    DISABLE_STREAMING_TESTS or get_package_version("hdmf") >= version.parse("3.3.1"),
     reason="Needs h5py setup with ROS3, as well as 'hdmf<3.3.1'.",
 )
 def test_check_none_matnwb_resolution_pass():

